### DISCUSSION TITLE
The ability to manage child components from the parent and global components

### DISCUSSION THREAD

**Discussion by dblackCat** [alpinejs/alpine#4236] [3. Feature ideas]
Hello everyone. I have been using alpine for several years and throughout my work with it, I constantly had to write the same binds for different components.

For example, there is a form in which there is a preloader.

There is a modal window and it has a preloader.

There is an online store basket, catalog filters and there is also a preloader.

And so I have binding in a large number of components.

```js
bindPreloader: {
     [':class']() {
         return (this.isProcess) ? '--active' : '';
     }
}
```

And this is just one of the situations where the code is duplicated. There are also various alerts, overlays and other elements. Which should be put into separate components in a good way. But in this case, it will be impossible to manage them.

Initially, I tried to cheat and throw a link from the child component to the parent one. And what was my surprise when it worked fine.

After a while, I decided to simplify the development process a bit and developed 2 plugins for alpine.js that I've been using for the last month.

Here is a link to my plugins: https://github.com/dblackCat/alpine_toolkit

### 1. xGlobal

Allows you to make any component global. After declaring the x-global="alias" directive, the component is added to the Components global variable and can be called from any js code.

**Example:**

```html
<div x-data="alert" x-global="globalAlert">alert text</div>
```

Inside any component or regular js code, you can access the global component.

```js
Components.globalAlert.show();
```

### 2. xSocket

Allows you to explicitly "connect" one component to another. In fact, the socket component will have an array with links to the connected components.

**Example:**

```html
<div x-data="form" x-socket="form">
    <div x-data="alert" x-connect.form="alert">alert text</div>
</div>

<div x-data="preloader" x-connect.form="preloader"></div>
```

Inside the form component, you can access all connected:

```js
this.connected.alert.show();
this.connected.preloader.show();
```

I repeat once again. It's amazing that it all works. And it works quite well.

I've only noticed one problem right now. For example, the components connected to the socket must be lower than the socket itself.

This is because the initialization of the components is performed as the page loads.

At the same time, global components do not cause an error. But if you use them in the init() method of the plugin, which is loaded earlier on the DOM tree, an error will occur.

I am writing this post to show that alpine js can be used more flexibly. And I would be extremely happy if such features were built into the framework.

I am a backend developer and I cannot call myself a great javascript professional, but I would like to discuss such improvements. 

If you don't like my ideas, then maybe someone will tell me how to improve my plugins in order to avoid mistakes when using them together in initialization methods.

**Comment by ekwoka** [comment]
These can be done with a custom plugin.

But also, I don't see why you need to rewrite that binding over and over?

It will work find just being in a normal binding and will reference the state of the component using it.

**Comment by dblackCat** [comment]
Perhaps I explained the situation incorrectly. I'll try to add some examples.

Without my plugins, I used to do this:

**Component: productForm**

```js
/**
 * Form component
 */
export default (action) => ({
    isProcess: false,
    message: null,
    isMessageSuccess: false,
    fields: {},

   /**
   * Processing and sending the form
   */
    send() {
        this.isProcess = true;

        // The logic of sending the form... 

        this.isProcess = false;
        this.message = response.message;
        this.isMessageSuccess = response.status;
    },


   /**
   * Disabling fields during an active process
   */
    bindField: {
        [':disabled']() {
            return this.isProcess;
        }
    },


   /**
   * Showing the loading animation while the process is active
   */
    bindProcess: {
        [':class']() {
            return (this.isProcess) ? '--active' : '';
        },
    },


   /**
   * The logic of the alert. Show, hide and style management
   */
    bindMessage: {
        [':class']() {
            let classList = (this.isMessageSuccess) ? 'alert--success' : '';
            classList += (this.message) ? '--active' : '';
          
            return classList;
        },
        ['x-text']() {
            return this.message;
        },
        ['@click']() {
            this.message = null;
        }
    }
});

```

```html
<!-- Form component markup -->
<form x-data="productForm('/my-handler')" @submit.prevent="send()">
    <input type="text" x-model.fill="fields.name" x-bind="bindField" placeholder="Product name">
    <input type="number" x-model.fill="fields.price" x-bind="bindField" placeholder="Price">

    <div class="preloader" x-bind="bindProcess">
         <!-- There's an svg animation here. The spinning circle -->
    </div>

    <div class="alert" x-bind="bindMessage"><div>
    
    <button type="submit">Send</button>
</form>
```

That's about how I wrote the components. It looks and works well. But there can be many such forms. 

Examples:

1. productForm
2. customerForm
3. InvoiceForm
4. and so on.

And each such form will have its own bindProcess and bindMessage, which on the one hand corresponds to the SRP principle, but completely contradicts DRY.

> These can be done with a custom plugin.

Using the link, I just threw off custom plugins that are trying to solve this problem. https://github.com/dblackCat/alpine_toolkit

> But also, I don't see why you need to rewrite that binding over and over?

And how do you solve this problem?

> It will work find just being in a normal binding and will reference the state of the component using it.

Yes, I understand what you mean. Here is an example:

```js
/**
 * Form component
 */
export default (action) => ({
    isProcess: false,
    message: null,
    isMessageSuccess: false,
    fields: {},

   /**
   * Processing and sending the form
   */
    send() {
        this.isProcess = true;

        // The logic of sending the form... 

        this.isProcess = false;
        this.message = response.message;
        this.isMessageSuccess = response.status;
    },


   /**
   * Disabling fields during an active process
   */
    bindField: {
        [':disabled']() {
            return this.isProcess;
        }
    }
});


/**
 * Alert component
 */
export default () => ({
   /**
   * The logic of the alert. Show, hide and style management
   */
    bindAlert: {
        [':class']() {
            let classList = (this.isMessageSuccess) ? 'alert--success' : '';
            classList += (this.message) ? '--active' : '';
          
            return classList;
        },
        ['x-text']() {
            return this.message;
        },
        ['@click']() {
            this.message = null;
        }
    }
});


/**
 * Preloader component
 */
export default () => ({
   /**
   * Showing the loading animation while the process is active
   */
    bindPreloader: {
        [':class']() {
            return (this.isProcess) ? '--active' : '';
        },
    },
});
```

```html
<!-- Form component markup -->
<form x-data="productForm('/my-handler')" @submit.prevent="send()">
    <input type="text" x-model.fill="fields.name" x-bind="bindField" placeholder="Product name">
    <input type="number" x-model.fill="fields.price" x-bind="bindField" placeholder="Price">

    <div class="preloader" x-bind="bindPreloader">
         <!-- There's an svg animation here. The spinning circle -->
    </div>

    <div class="alert" x-bind="bindAlert"><div>
   
    <button type="submit">Send</button>
</form>
```

It will work. But there are a couple of big disadvantages.

1. Implicit usage. With this usage, there should be only one alert in the form. And if this is not critical for the current example, then there are completely different examples where this is required.
2. This usage completely eliminates the possibility of using the preloader and alert components separately from the parent. Again, this is acceptable for this example. But the components are different.

And again, let's touch on implicit usage. If we have several nested components, it becomes more difficult to determine where an alert is controlled from, for example.

I hope I was able to explain the situation better. In the next post, I will give a detailed example of usage through my plugins.

**Comment by dblackCat** [comment]
Now I'll show you what I've come to. Also with the help of detailed examples.

This example uses the xSocket plugin: https://github.com/dblackCat/alpine_toolkit/tree/main/src/x-socket

```js
/**
 * ProductForm component
 */
export default (action) => ({
    isProcess: false,
    fields: {},

   /**
   * Processing and sending the form
   */
    send() {
        this.isProcess = true;
        this.connected.preloader.show();

        // The logic of sending the form... 

        this.isProcess = false;
        this.connected.preloader.hide();
        this.connected.alertOne.show(response.message, response.status);
        this.connected.alertTwo.show("This can also be used. Although it doesn't make sense in this example either.", response.status);
    },


   /**
   * Disabling fields during an active process
   */
    bindField: {
        [':disabled']() {
            return this.isProcess;
        }
    },
});


/**
 * Preloader component
 */
export default () => ({
    isShow: false,

    /**
     * Show
     */
    show() {
        this.isShow = true;
    },

    /**
     * Hide
     */
    hide() {
        this.isShow = false;
    },

    /**
     * Bind preloader
     */
    bindPreloader: {
        [':class']() {
            return (this.isShow) ? '--active' : '';
        },
    },
});


/**
 * Alert component
 */
export default () => ({
    isSuccess: false,
    type: null,

    /**
     * Show alert
     */
    show(message, isSuccess = false) {
        this.message = message;
        this.isSuccess = isSuccess;
    },

    /**
     * Hide alert
     */
    hide() {
        this.message = null;
        this.isSuccess = false;
    },

    /**
     * Bind alert
     */
    bindAlert: {
        [':class']() {
            let classList = (this.isSuccess) ? 'alert--success' : '';
            classList += (this.message) ? '--active' : '';
          
            return classList;
        },
        ['x-text']() {
            return this.message;
        },
        ['@click']() {
            this.hide();
        }
    }
});

```

```html
<!-- Form component markup -->
<form x-data="productForm('/my-handler')" @submit.prevent="send()" x-socket="product_form">
    <input type="text" x-model.fill="fields.name" x-bind="bindField" placeholder="Product name">
    <input type="number" x-model.fill="fields.price" x-bind="bindField" placeholder="Price">

    <div class="preloader" x-data="preloader" x-bind="bindPreloader" x-connect.product_form="preloader">
         <!-- There's an svg animation here. The spinning circle -->
    </div>

    <div class="alert" x-data="alert" x-bind="bindAlert" x-connect.product_form="alertOne"><div>
    <div class="alert" x-data="alert" x-bind="bindAlert" x-connect.product_form="alertTwo"><div>
    
    <button type="submit">Send</button>
</form>
```
In this example, we already have explicit use of components. At the same time, the components remain independent and can be used separately.

This is not relevant for the preloader and alert components, they are usually always dependent. But there are other components that you want to use both together and separately from each other.

In this example, we didn't have to write duplicate bindings. Each component determines them by itself, which seems to me to be correct.

**Reply by ekwoka** [reply]
So, I'd mainly point to this as a design issue, that you're thinking about imperative control instead of declarative control.

[here is an example of doing it more declaratively](https://awesomealpine.com/play/?coreplugins=0&html=PGZvcm0KICBjbGFzcz0icmVsYXRpdmUgZmxleCBmbGV4LWNvbCBnYXAtMyIKICB4LWRhdGE9InByb2R1Y3RGb3JtKCcvbXktaGFuZGxlcicpIgogIEBzdWJtaXQucHJldmVudD0ic3VibWl0KCRlbCkiCj4KICA8aW5wdXQKICAgIHR5cGU9InRleHQiCiAgICBuYW1lPSJuYW1lIgogICAgeC1tb2RlbC5maWxsPSJmaWVsZHMubmFtZSIKICAgIHgtYmluZD0iZm9ybUZpZWxkIgogICAgcGxhY2Vob2xkZXI9IlByb2R1Y3QgbmFtZSIKICAvPgogIDxpbnB1dAogICAgdHlwZT0ibnVtYmVyIgogICAgbmFtZT0icHJpY2UiCiAgICB4LW1vZGVsLmZpbGw9ImZpZWxkcy5wcmljZSIKICAgIHgtYmluZD0iZm9ybUZpZWxkIgogICAgcGxhY2Vob2xkZXI9IlByaWNlIgogIC8-CgogIDxkaXYKICAgIHgtY2xvYWsKICAgIGNsYXNzPSJhYnNvbHV0ZSBpbnNldC0wIGJnLWdyYXktMzAwLzUwIHctZnVsbCBoLWZ1bGwgei0xMCIKICAgIHgtc2hvdz0ic2VuZGluZyIKICA-CiAgICA8IS0tIFRoZXJlJ3MgYW4gc3ZnIGFuaW1hdGlvbiBoZXJlLiBUaGUgc3Bpbm5pbmcgY2lyY2xlIC0tPgogIDwvZGl2PgoKICA8dGVtcGxhdGUKICAgIHgtZm9yPSJba2V5LCBtZXNzYWdlXSBvZiBPYmplY3QuZW50cmllcyhtZXNzYWdlcykiCiAgICA6a2V5PSJrZXkiCiAgPgogICAgPGRpdgogICAgICBjbGFzcz0iYWxlcnQiCiAgICAgIHgtdGV4dD0ibWVzc2FnZSIKICAgID48L2RpdgogID48L3RlbXBsYXRlPgoKICA8YnV0dG9uCiAgICB0eXBlPSJzdWJtaXQiCiAgICB4LXRleHQ9InNlbmRpbmcgPyAnc2VuZGluZycgOiAnU2VuZCciCiAgICBAY2xpY2sucHJldmVudD0ic3VibWl0KCRlbC5mb3JtKSIKICA-CiAgICBTZW5kCiAgPC9idXR0b24-CjwvZm9ybT4K&tw=1&ts=1&script=QWxwaW5lLmRhdGEoJ3Byb2R1Y3RGb3JtJywgKF9oYW5kbGVyKSA9PiAoewogIHNlbmRpbmc6IGZhbHNlLAogIG1lc3NhZ2VzOiB7fSwKICBmaWVsZHM6IHt9LAogIGFzeW5jIHN1Ym1pdChmb3JtOiBIVE1MRm9ybUVsZW1lbnQpIHsKICAgIGNvbnN0IF9kYXRhID0gbmV3IEZvcm1EYXRhKGZvcm0pOwogICAgdGhpcy5zZW5kaW5nID0gdHJ1ZTsKICAgIGF3YWl0IG5ldyBQcm9taXNlKChyZXMpID0-IHNldFRpbWVvdXQocmVzLCAxMDAwKSk7CiAgICB0aGlzLm1lc3NhZ2VzID0geyBwcmljZTogJ1ByaWNlIG11c3QgYmUgYmVsb3cgJDEwMCcgfTsKICAgIHRoaXMuc2VuZGluZyA9IGZhbHNlOwogIH0sCn0pKTsKCkFscGluZS5iaW5kKCdmb3JtRmllbGQnLCB7CiAgJzpkaXNhYmxlZCcoKSB7CiAgICByZXR1cm4gdGhpcy5zZW5kaW5nOwogIH0sCn0pOwo)

You just bind on to data. You don't tell things to do things, but instead let the data naturally cascade.
